### DISCUSSION TITLE
Feature: Add timeout option

### DISCUSSION THREAD

**Discussion by proost** [redis/rueidis#468] [Ideas]
When use redis with rueidis, i set timeout to `context.Context`. but rueidis can handle this, so that many rueidis user do not have to set timeout to `context.Context`.

[`http.Client`](https://pkg.go.dev/net/http#Client) supports timeout, and [lettuce](https://lettuce.io/core/release/reference/#client-options) in Java supports timeout. if rueidis support timeout option, then it can be more convenient.

**Comment by rueian** [comment]
Hi @proost, rueidis does have a default timeout option: `option.ConnWriteTimeout`. It is 10s by default.

**Reply by proost** [reply]
Oh, I see. after i dig out code, i know that `option.ConnWriteTimeout` can be used as request timeout. comments above `ConnWriteTimeout` quite confused to me. Because comments on `ConnWriteTimeout` looks like "only" using for keeping alive connection with `ClientOption.Dialer.KeepAlive`. Adding explicitly a comment that `ConnWriteTimeout` can be used as a request timeout is more helpful like me, who english isn't native language. what do you think about it?

Also, i found that if requested context has deadline, only use deadline from requested context. When `ConnWriteTimeout` is configured and requested context has deadline, setting shorter deadline between them to connection is reasonable to me. are there reasons using one of them?

**Reply by rueian** [reply]
1. Sure, mentioning that `ConnWriteTimeout` can be used as timeout is great.
2. There is no special reason for that. Setting the shorter one looks good to me.

**Reply by proost** [reply]
There is something confused me.

```
	if dlOk {
		p.conn.SetDeadline(dl)
	} else if p.timeout > 0 && !cmd.IsBlock() {
		p.conn.SetDeadline(time.Now().Add(p.timeout))
	} else {
		p.conn.SetDeadline(time.Time{})
	}
```

setting timeout logic likes above,
1. if a `context.Context` has deadline, whether some commands are blocking or not, always set timeout.
2. if a `context.Context` has not deadline, `ConnWriteTimeout` is configured and all commands is not block, set timeout.

I think, these logic feels like inconsistent. Because if one of commands is blocking, always ignore `pipe.timeout` and deadline in `context.Context` or respect `pipe.timeout` and deadline in `context.Context` both. why respect only deadline in `context.Context`? what are reasons?

**Reply by rueian** [reply]
Consider the following two use cases:

```go
func case1() {
	client, _ := rueidis.NewClient(rueidis.ClientOption{InitAddress: []string{"127.0.0.1:6379"}})
	defer client.Close()

	client.Do(context.Background(), client.B().Blpop().Key("list").Timeout(0).Build())
}

func case2() {
	client, _ := rueidis.NewClient(rueidis.ClientOption{InitAddress: []string{"127.0.0.1:6379"}})
	defer client.Close()

	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
	defer cancel()
	client.Do(ctx, client.B().Blpop().Key("list").Timeout(0).Build())
}
```

Users who write the case1 will want the BLPOP block indefinitely, therefore no deadline is applied.

Users who write the case2 will want the BLPOP block for at most 1 minute, therefore 1 minute deadline is applied.

**Reply by proost** [reply]
Yes, I agree. we should avoid block indefinitely. so if `pipe.timeout` is configured, no deadline in `context.Context` and blocking command is given, we should set deadline too.

**Reply by rueian** [reply]
Uh, no, that is not what I mean.

I mean users want to block indefinitely in the first case, so we should not apply `pipe.timeout`.

**Reply by proost** [reply]
Oh, I understand.

**Comment by proost** [comment]
Solved by:
https://github.com/redis/rueidis/pull/470
https://github.com/redis/rueidis/pull/469
